utility/network.go

```diff
@@ -0,0 +1,118 @@
+package utility
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/bitrise-io/go-utils/fileutil"
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/bitrise-io/go-utils/pathutil"
+	"github.com/bitrise-io/go-utils/retry"
+
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+
+	"google.golang.org/api/androidpublisher/v3"
+)
+
+// CreateHTTPClient creates an HTTP client for the communication during the uploads.
+func CreateHTTPClient(jsonKeyPth string) (*http.Client, error) {
+	jsonKeyPth, isRemote, err := ParseURI(string(jsonKeyPth))
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare key path (%s), error: %s", jsonKeyPth, err)
+	}
+
+	if isRemote {
+		tmpDir, err := pathutil.NormalizedOSTempDirPath("__google-play-deploy__")
+		if err != nil {
+			return nil, fmt.Errorf("failed to create tmp dir, error: %s", err)
+		}
+
+		jsonKeySource := jsonKeyPth
+		jsonKeyPth = filepath.Join(tmpDir, "key.json")
+		if err := downloadFileWithRetry(jsonKeySource, jsonKeyPth, 3, 3); err != nil {
+			return nil, fmt.Errorf("failed to download json key file, error: %s", err)
+		}
+	}
+
+	authConfig, err := jwtConfigFromJSONKeyFile(jsonKeyPth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create auth config from json key file %v, error: %s", jsonKeyPth, err)
+	}
+	jwtConfig := authConfig
+
+	return jwtConfig.Client(context.TODO()), nil
+}
+
+// jwtConfigFromJSONKeyFile gets the jwt config from the given file.
+func jwtConfigFromJSONKeyFile(pth string) (*jwt.Config, error) {
+	jsonKeyBytes, err := fileutil.ReadBytesFromFile(pth)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg, err := google.JWTConfigFromJSON(jsonKeyBytes, androidpublisher.AndroidpublisherScope)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// ParseURI parses the given URI to return the path from it if it is a local file, if it is remote the bool value is
+// true.
+func ParseURI(keyURI string) (string, bool, error) {
+	jsonURL, err := url.Parse(keyURI)
+	if err != nil {
+		return "", false, fmt.Errorf("failed to parse url (%s), error: %s", keyURI, err)
+	}
+
+	return strings.TrimPrefix(keyURI, "file://"), jsonURL.Scheme == "http" || jsonURL.Scheme == "https", nil
+}
+
+// downloadFileWithRetry calls downloadFile method with a given number of retries and waiting interval between the retries.
+func downloadFileWithRetry(downloadURL, targetPath string, numberOfRetries, waitInterval uint) error {
+	if err := retry.Times(numberOfRetries).Wait(time.Duration(waitInterval) * time.Second).Try(func(attempt uint) error {
+		return downloadFile(downloadURL, targetPath)
```

Comment:
NOTE: would like to create standardized retry, as for example status code not found does not need retry IMO

Reply:
Agree with you, but this should be a general utility method (and should not be placed here), but this time I would say it is out of scope, as initially this was a bugfix.

Reply:
Yes, meant as a general comment, we agreed that the note prefix means that.